java/src/main/java/org/whispersystems/libsignal/util/Base32.java

```diff
@@ -1,157 +1,394 @@
 package org.whispersystems.libsignal.util;
 
-import java.util.HashMap;
-import java.util.Map;
-
 /**
- * An implementation of Base32 encoding/decoding that accepts arbitrary alphabets.
- * From http://www.herongyang.com/Encoding/Base32-Bitpedia-Java-Implementation.html;
- *
- * It also supports a replacement table on decoding to handle mis-entered letters.
+ * This is a port of https://github.com/paragonie/constant_time_encoding/blob/master/src/Base32.php
+ * intended to avoid cache timing attacks by avoiding the use of table lookups and branching
+ * constructs. The encoding and decoding logic only use bitwise operators except for length
+ * dependent logic, meaning that for inputs of a given length, encode/decode should be
+ * constant-time.
+ * <p>
+ * This does not use padding and uses only lower case.
+ * <p>
+ * This uses the alphabet ybndrfg8ejkmcpqxot1uw2sza345h769, which is similar to the z-base-32
+ * alphabet that preferences characters that are easier for humans to read, but our alphabet omits
+ * the number 0 and the letters i, l and v. On decoding, it also maps the letters i and l to the
+ * number 1 and the number 0 to the letter o.
  */
 public class Base32 {
-    private final String alphabet;
-
-    private static final int[] base32Lookup = new int[128];
-    private static final int[] replacementLookup = new int[128];
+    /**
+     * Decode from a Base32 ASCII array of bytes.
+     */
+    public static byte[] decode(byte[] src) {
+        return doDecode(new ByteArrayCharacterSource(src));
```

Comment:
NIT

```suggestion
        if (src == null) { return null; }
        if (len(src) == 0) { return null; }
        return doDecode(new ByteArrayCharacterSource(src));
```

The 2nd conditional is in doDecode(), but it's nice to clearly specify the expected input/output from the function's first few lines, rather than having the developer go inside the code.

Reply:
That's actually a change in the logic, the method will always return at least an empty byte[]. I'll add that to the comment.